src/include/detail/graph/nn-descent.h

```diff
@@ -0,0 +1,470 @@
+
+  g.build_in_edges();
+
+  for (size_t i = 0; i < nvertices; ++i) {
+    for (auto&& [_, j] : out_edges(g, i)) {
+      // print_types(i, j);
+      num_updates += nn_descent_1_step(g, db, i, j);
+    }
+
+    for (auto&& j : in_edges(g, i)) {
+      num_updates += nn_descent_1_step(g, db, i, j);
+    }
+  }
+  // g.swap_all_update_edges();
+
+  return num_updates;
+}
+
+#if 0
+template <adjacency_list_graph Graph>
+auto bfs(const Graph& graph, vertex_id_t<Graph> root) {
+  using vertex_id_type = vertex_id_t<Graph>;
+
+  std::deque<vertex_id_type>  q1, q2;
+  std::vector<vertex_id_type> level(num_vertices(graph), std::numeric_limits<vertex_id_type>::max());
+  std::vector<vertex_id_type> parents(num_vertices(graph), std::numeric_limits<vertex_id_type>::max());
+  size_t                      lvl = 0;
+
+  q1.push_back(root);
+  level[root]   = lvl++;
+  parents[root] = root;
+
+  while (!q1.empty()) {
+
+    std::for_each(q1.begin(), q1.end(), [&](vertex_id_type u) {
+      std::for_each(graph[u].begin(), graph[u].end(), [&](auto&& x) {
+        vertex_id_type v = target(graph, x);
+        if (level[v] == std::numeric_limits<vertex_id_type>::max()) {
+          q2.push_back(v);
+          level[v]   = lvl;
+          parents[v] = u;
+        }
+      });
+    });
+    std::swap(q1, q2);
+    q2.clear();
+    ++lvl;
+  }
+  return parents;
+}
+
+#endif
+
+template <class Graph>
+struct vertex_id {
+  using type = typename Graph::vertex_id_type;
+};
+
+template <class Graph>
+using vertex_id_t = typename vertex_id<Graph>::type;
+
+template <class T, class I>
+struct vertex_id<nn_graph<T, I>> {
+  using type = I;
+};
+
+template <class T, class I>
+auto bfs(nn_graph<T, I>& graph, I root) {
+  using vertex_id_type = I;
+
+  std::deque<vertex_id_type> q1, q2;
+  std::vector<vertex_id_type> level(
+      num_vertices(graph), std::numeric_limits<vertex_id_type>::max());
+  std::vector<vertex_id_type> parents(
+      num_vertices(graph), std::numeric_limits<vertex_id_type>::max());
+  size_t lvl = 0;
+
+  q1.push_back(root);
+  level[root] = lvl++;
+  parents[root] = root;
+
+  while (!q1.empty()) {
+    std::for_each(q1.begin(), q1.end(), [&](vertex_id_type u) {
+      std::for_each(
+          graph.out_edges(u).begin(), graph.out_edges(u).end(), [&](auto&& x) {
+            auto&& [_, v] = x;
+            if (level[v] == std::numeric_limits<vertex_id_type>::max()) {
+              q2.push_back(v);
+              level[v] = lvl;
+              parents[v] = u;
+            }
+          });
+    });
+    std::swap(q1, q2);
+    q2.clear();
+    ++lvl;
+  }
+
+  for (auto&& l : level) {
+    if (l == std::numeric_limits<vertex_id_type>::max()) {
+      std::cout << "Graph is not connected!" << std::endl;
+      break;
+    }
+  }
+
+  return parents;
+}
+
+template <class T, std::integral ID, class Distance = sum_of_squares_distance>
+auto nn_descent_1_query(
+    const detail::graph::nn_graph<T, ID>& graph,
+    auto&& db,
```

Comment:
Having arguments specified as `auto` greatly reduces the readability of this code